### ISSUE TITLE
raw.read() unexpected behavior with gzip compression - needs docs

### ISSUE THREAD

**Issue by foxx** [urllib3/urllib3#189]
Hello,

Discussed briefly here;
https://github.com/kennethreitz/requests/issues/1401#issuecomment-18829272

If you use .raw.read() on a gzip compressed request, it returns a bunch of unicode because of the compression. By default, I think it should use 'decode_content=True', as this is again a bit confusing. I can't think of any situation where we would want to read in non decoded data, or at the very least the majority of situations would not require this.

Thoughts??

Cal

**Comment by seyhuns** [comment]
@foxx I am experiencing the same.

Besides returning empty bytes, another problem is that ```read``` can return multiple bytes:
```python
>>> http_response
<urllib3.response.HTTPResponse object at 0x10ce53350>
>>> http_response.headers['Content-Encoding']
'gzip'
>>> http_response.read(1, decode_content=True)
b''
>>> http_response.read(1, decode_content=True)
b'd'
>>> http_response.read(1, decode_content=True)
b'8'
>>> http_response.read(1, decode_content=True)
b':a'
```

I am planning to wrap my response with [GzipFile](https://docs.python.org/3/library/gzip.html#gzip.GzipFile) to solve this problem.

```python
>>> http_response
<urllib3.response.HTTPResponse object at 0x109d06490>
>>> http_response.getheader('Content-Encoding')
'gzip'
>>> wrapped_http_response = gzip.GzipFile(fileobj=http_response)
>>> wrapped_http_response.read(1)
b'd'
>>> wrapped_http_response.read(1)
b'8'
>>> wrapped_http_response.read(1)
b':'
>>> wrapped_http_response.read(1)
b'a'
```

I believe this issue should remain open. In my example, I am explicitly asking for `amt=1` and `read` is returning bytes with different lengths.

**Comment by sigmavirus24** [comment]
This should remain closed. It's been closed for 7 years. 2 people in approximately that time frame isn't a reason to keep an issue open indefinitely.

**Comment by seyhuns** [comment]
Thank you for checking, @sigmavirus24.

What happens if I try to find a solution? Can I contribute? Or the fix won't be applied because this issue is closed?

I don't know even if I can find a solution or not, but just wondering and trying to understand the workflow.

Again, thank you a lot for your time.

**Comment by sigmavirus24** [comment]
> What happens if I try to find a solution?

You may never use urllib3 again if you try to find a solution. ;) (By which I'm jokingly trying to say you'll hate the experience so much you'll never want to look at the name "urllib3" again. :D)

> Can I contribute? Or the fix won't be applied because this issue is closed?

If you manage to find a backwards-compatible fix that doesn't change the semantics of `.read()` for all of its potential use-cases, then yes, I expect that contribution would be accepted. As someone who has made several attempts to do just that, I would be happy to review your contribution.

> I don't know even if I can find a solution or not, but just wondering and trying to understand the workflow.

Contributions aren't based on open bugs strictly (to the best of my knowledge).

---

Let me give you a word of warning though: The end user behaviour here is surprising (to the very few people who use it), _but_ the internals that dictate that behaviour are straight-forward and easy to understand. That's a large part of why things haven't changed. It's easier to maintain something with an odd end-user behaviour than to try to make it fit what the end-user _might_ expect. Specifically, your `read` calls aren't reading 1 byte off of the `GzipFile` but instead are reading 1 byte off of the socket from the server and passing it to the GzipFile to transform. That's why the returned byte objects are not of uniform length. We would need to change how we deal with the underlying socket based off of a parameter passed to each call. And what happens if someone changes the value of `decode_content` between two calls? What behaviour do we do then? If someone said `Give me one decoded byte` we'd have to read enough from the socket to do so, but in doing so, we will have _probably_ read several bytes from the socket. If someone then says `Give me one byte` what byte do we give them? We may have fed the decoder more than what was necessary to get the next decompressed byte, but then semantically are we giving the user the next byte off of the socket? From the gzip decoder? What? 

In short, the design of this particular part of the API is flawed. That flaw makes it incredibly difficult to reason about how to "fix" it.

**Comment by seyhuns** [comment]
Thanks for sharing your experience with this. It was very enlightening to me.
